lib/thrift/generator/client.ex

```diff
@@ -0,0 +1,274 @@
+                     message = "Both sequence id and rpc name are wrong. The server is extremely uncompliant."
+                     %TApplicationException{message: message,
+                                            type: :sequence_id_and_rpc_name_mismatched}
+                 end
+            {:error, {:exception, ex}}
+          end
+
+          defp handle_message({:error, _} = err, _, _, _) do
+            err
+          end
+
+          defp read_t_application_exception(
+                <<11::size(8),
+                1::16-unsigned,
+                message_size::32-signed,
+                message::binary-size(message_size),
+                rest::binary>>, accum) do
+            # read the message string
+            read_t_application_exception(rest, Map.put(accum, :message, message))
+          end
+          defp read_t_application_exception(
+                <<8::size(8),
+                2::16-unsigned,
+                type::32-signed,
+                rest::binary>>, accum) do
+            # read the type
+            exception_type = TApplicationException.exception_type(type)
+            read_t_application_exception(rest, Map.put(accum, :type, exception_type))
+          end
+          defp read_t_application_exception(<<0>>, accum) do
+            # read the field stop and return
+            accum
+          end
+          defp read_t_application_exception(error, _) do
+            message = "Could not decode TApplicationException, remaining was #{inspect error}"
+            %TApplicationException{message: message,
+                                   type: TApplicationException.exception_type(7)}
+          end
+        end
+      end
+    end
+
+    def generate_function(service_module, function) do
+
+      args_module = service_module
+      |> Module.concat(Service.service_module_name(function, :args))
+
+      response_module = Service.service_module_name(function, :response)
+
+      underscored_name = function.name
+      |> Atom.to_string
+      |> Macro.underscore
+      |> String.to_atom
+
+      bang_name = :"#{underscored_name}!"
+
+      vars = function.params
+      |> Enum.map(&Macro.var(&1.name, nil))
+
+      assignments = function.params
+      |> Enum.zip(vars)
+      |> Enum.map(fn {param, var} ->
+        quote do
+          {unquote(param.name), unquote(var)}
+        end
+      end)
+
+      rpc_name = Atom.to_string(function.name)
+
+      quote do
+        def unquote(underscored_name)(client, unquote_splicing(vars), timeout \\ 5000) do
+          serialized_args = %unquote(args_module){unquote_splicing(assignments)}
+          |> unquote(args_module).BinaryProtocol.serialize
+
+          sequence_id = :erlang.unique_integer([:positive])
+          message = Binary.serialize(
+            :message_begin,
+            {unquote(message_type(function)), sequence_id, unquote(rpc_name)})
+
+          GenServer.call(client, {:send, [message | serialized_args]})
```

Comment:
Using the two `GenServer` calls requires a lock on the process and unsure if lock is guaranteed or not. It would also be more efficient to do a single call but may not make significant difference as the server has to receive and handle the call.